Index: linux-3.13.7/drivers/reset/Kconfig
===================================================================
--- linux-3.13.7.orig/drivers/reset/Kconfig
+++ linux-3.13.7/drivers/reset/Kconfig
@@ -11,3 +11,7 @@ menuconfig RESET_CONTROLLER
 	  via GPIOs or SoC-internal reset controller modules.
 
 	  If unsure, say no.
+
+config RESET_CONTROLLER_OXNAS
+	bool
+	select RESET_CONTROLLER
Index: linux-3.13.7/drivers/reset/Makefile
===================================================================
--- linux-3.13.7.orig/drivers/reset/Makefile
+++ linux-3.13.7/drivers/reset/Makefile
@@ -1 +1,2 @@
 obj-$(CONFIG_RESET_CONTROLLER) += core.o
+obj-$(CONFIG_RESET_CONTROLLER_OXNAS) += reset-ox820.o
